### PR TITLE
Add key predicate to baggage span processor

### LINKED ISSUE #2472: Baggage span processor - key predicate
This issue is to track adding a method of selecting what baggage key entries should be copied.

Feedback in the JS contrib PR was to allow a user-provided predicate function. This puts the responsibility on the user to ensure sensitive baggage keys are not copied while also not prescribing how that is determined.
- https://github.com/open-telemetry/opentelemetry-js-contrib/issues/2166

We had a similar feedback in the .NET contrib project but thought it was more complicated than just using a set of prefixes so created an issue to continue the discussion. The plain processor that copies all baggage entries (like using `*` in your example) is likely to be accepted first.
- https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/1695

### PR DESCRIPTION
**Pull request by MikeGoldsmith**
# Description

Updates the Baggage span processor to require a baggage key predicate function that selects what keys are copied to new spans.

Fixes #2472 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
  (The original span processor PR #2428 has been published yet so this is not breaking)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added unit tests

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: processor/opentelemetry-processor-baggage/README.rst [hunk 1]
```diff
@@ -20,3 +20,32 @@ Do not put sensitive information in Baggage.
...
+
+Add the span processor when configuring the tracer provider.
+
+To configure the span processor to copy all baggage entries during configuration:
+
+```python
+from opentelemetry.processor.baggage import BaggageSpanProcessor, ALLOW_ALL_BAGGAGE_KEYS
+
+tracer_provider = TracerProvider()
+tracer_provider.add_span_processor(BaggageSpanProcessor(ALLOW_ALL_BAGGAGE_KEYS))
```
Comment (codeboten) [inline processor/opentelemetry-processor-baggage/README.rst:34]: Does this need to be a required arg or could this just be the default value?

File: processor/opentelemetry-processor-baggage/tests/test_baggage_processor.py [hunk 2]
```diff
@@ -59,7 +127,9 @@ def test_set_baggage_attaches_to_child_spans_and_detaches_properly_with_token(
         self,
     ):
         tracer_provider = TracerProvider()
-        tracer_provider.add_span_processor(BaggageSpanProcessor())
+        tracer_provider.add_span_processor(
+            BaggageSpanProcessor(ALLOW_ALL_BAGGAGE_KEYS)
```
Comment (codeboten) [inline processor/opentelemetry-processor-baggage/tests/test_baggage_processor.py:131]: If this is a required argument, then it should be logged as a breaking change

File: processor/opentelemetry-processor-baggage/README.rst (hunk 1 above)
Comment (MikeGoldsmith) [reply processor/opentelemetry-processor-baggage/README.rst:34]: For consistency with other SDK implementations, I went went with a required arg to help make the processor secure by default by requiring the developer to set the allows baggage keys. If it were a default parameter, it's easy to add the processor and not know all keys are being copied.

File: processor/opentelemetry-processor-baggage/tests/test_baggage_processor.py (hunk 2 above)
Comment (MikeGoldsmith) [reply processor/opentelemetry-processor-baggage/tests/test_baggage_processor.py:131]: Yeah, it is a breaking change in that we've introduced a required parameter. However, this processor hasn't been released yet so was trying to lean on not declaring a breaking change 😄

Comment (codeboten) [reply processor/opentelemetry-processor-baggage/tests/test_baggage_processor.py:131]: I guess so long as it's merged before the release, this should be ok

**Review by codeboten — APPROVED** [review]

**Review by ocelotl — APPROVED** [review]

**Pull request merged**